pkg/filewatcher/filewatcher.go

```diff
@@ -0,0 +1,253 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filewatcher
+
+import (
+	"bufio"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/hashicorp/go-multierror"
+)
+
+type FileWatcher interface {
+	Add(path string) error
+	Remove(path string) error
+	Close() error
+	Events(path string) chan fsnotify.Event
+	Errors(path string) chan error
+}
+
+type fsNotifyWatcher struct {
+	mu sync.RWMutex
+	// The watcher maintain a map of workers,
+	// keyed by watched dir (parent dir of watched files).
+	workers map[string]*worker
+}
+
+type worker struct {
+	// watcher is a fsnotify watcher that watches the parent
+	// dir of watchedFiles.
+	watcher *fsnotify.Watcher
+	// The worker maintain a map of event channel (included in fileTracker),
+	// keyed by watched file path.
+	// The worker watches parent path of given path,
+	// and filter out events of given path, then redirect
+	// to the result channel.
+	// Note that for symlink files, the content in received events
+	// do not have to be related to the file itself.
+	watchedFiles map[string]*fileTracker
+}
+
+type fileTracker struct {
+	events chan fsnotify.Event
+	// md5 sum to indicate if a file has been updated.
+	md5Sum string
+}
+
+// NewWatcher return with a FileWatcher instance that implemented with fsnotify.
+func NewWatcher() FileWatcher {
+	w := &fsNotifyWatcher{
+		workers: map[string]*worker{},
+	}
+
+	return w
+}
+
+// Add is implementation of Add interface of FileWatcher.
+func (w *fsNotifyWatcher) Add(path string) error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	cleanedPath, parentPath := formatPath(path)
+	if worker, workerExist := w.workers[parentPath]; workerExist {
+		if _, pathExist := worker.watchedFiles[cleanedPath]; pathExist {
+			return fmt.Errorf("path %s already added", path)
+		}
+
+		// And the path into worker maps if not exist.
+		md5Sum, err := getMd5Sum(cleanedPath)
+		if err != nil {
+			return fmt.Errorf("failed to get md5 sum for %s: %v", path, err)
+		}
+		worker.watchedFiles[cleanedPath] = &fileTracker{
+			events: make(chan fsnotify.Event, 1),
+			md5Sum: md5Sum,
+		}
+
+		return nil
+	}
+
+	// Build a new worker if not exist.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	if err = watcher.Add(parentPath); err != nil {
+		return err
+	}
+
+	md5Sum, err := getMd5Sum(cleanedPath)
+	if err != nil {
+		return fmt.Errorf("failed to get md5 sum for %s: %v", path, err)
+	}
+	wk := &worker{
+		watcher: watcher,
+		watchedFiles: map[string]*fileTracker{
+			cleanedPath: {
+				events: make(chan fsnotify.Event, 1),
+				md5Sum: md5Sum,
+			},
+		},
+	}
+	w.workers[parentPath] = wk
+
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok { // 'Events' channel is closed
+					return
+				}
+
+				for path, tracker := range wk.watchedFiles {
+					newSum, _ := getMd5Sum(path)
```

Comment:
Wouldn't getting md5sum for each event be expensive? We will be re-md5summing all the files watched under the folder for a change in any one of them, right?

Would looking at other stat info be cheaper? (i.e. length, modified time, sym-link target change  etc.)

Reply:
From a practical perspective, this should be ok. We can keep it as is for the time being.